plugins/search-backend-module-elasticsearch/src/engines/ElasticSearchSearchEngine.ts

```diff
@@ -0,0 +1,265 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  IndexableDocument,
+  SearchQuery,
+  SearchResultSet,
+  SearchEngine,
+} from '@backstage/search-common';
+import { Logger } from 'winston';
+import esb from 'elastic-builder';
+import { Client } from '@elastic/elasticsearch';
+import { Config } from '@backstage/config';
+import {
+  createAWSConnection,
+  awsGetCredentials,
+} from '@acuris/aws-es-connection';
+import { isEmpty, isNaN as nan, isNumber } from 'lodash';
+
+export type ConcreteElasticSearchQuery = {
+  documentTypes?: string[];
+  elasticQueryBuilder: () => any;
+};
+
+type ElasticConfigAuth = {
+  username: string;
+  password: string;
+  apiKey: string;
+  bearer: string;
+};
+
+type ElasticSearchQueryTranslator = (
+  query: SearchQuery,
+) => ConcreteElasticSearchQuery;
+
+type ElasticSearchOptions = {
+  logger: Logger;
+  config: Config;
+  aliasPostfix?: string;
+  indexPrefix?: string;
+};
+
+type ElasticSearchResult = {
+  _index: string;
+  _type: string;
+  _score: number;
+  _source: IndexableDocument;
+};
+
+function duration(startTimestamp: [number, number]): string {
+  const delta = process.hrtime(startTimestamp);
+  const seconds = delta[0] + delta[1] / 1e9;
+  return `${seconds.toFixed(1)}s`;
+}
+
+function isBlank(str: string) {
+  return (isEmpty(str) && !isNumber(str)) || nan(str);
+}
+
+export class ElasticSearchSearchEngine implements SearchEngine {
+  constructor(
+    private readonly elasticSearchClient: Client,
+    private readonly aliasPostfix: string,
+    private readonly indexPrefix: string,
+    private readonly logger: Logger,
+  ) {}
+
+  static async initialize({
```

Comment:
What do you think of following the `Class.fromConfig(config, { dependencies })` pattern from elsewhere in backstage?

Reply:
Yup, makes sense :+1: